hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/debug/replicas/ReplicaStateVerifier.java

```diff
@@ -0,0 +1,222 @@
+    ReplicaKey key = new ReplicaKey(dn, containerId);
+    ContainerDataProto cachedData = containerDataCache.getIfPresent(key);
+    if (cachedData != null) {
+      return cachedData;
+    }
+    // Cache miss - fetch and store
+    ContainerDataProto data = fetchContainerDataFromDatanode(dn, containerId, keyLocation,
+        replicaIndex, containerInfoToken);
+    if (data != null) {
+      containerDataCache.put(key, data);
+    }
+    return data;
+  }
+
+  private ContainerDataProto fetchContainerDataFromDatanode(DatanodeDetails dn, long containerId,
+                                                            OmKeyLocationInfo keyLocation, int replicaIndex,
+                                                            ContainerInfoToken containerInfoToken)
+      throws IOException {
+    XceiverClientSpi client = null;
+    ReadContainerResponseProto response;
+    try {
+      Pipeline pipeline = Pipeline.newBuilder(keyLocation.getPipeline())
+          .setReplicationConfig(StandaloneReplicationConfig.getInstance(ONE))
+          .setNodes(Collections.singletonList(dn))
+          .setReplicaIndexes(Collections.singletonMap(dn, replicaIndex))
+          .build();
+      String encodedToken = containerInfoToken.getEncodedToken();
+
+      client = xceiverClientManager.acquireClientForReadData(pipeline);
+      response = ContainerProtocolCalls
+          .readContainer(client, containerId, encodedToken);
+    } finally {
+      if (client != null) {
+        xceiverClientManager.releaseClient(client, false);
+      }
+    }
+
+    if (!response.hasContainerData()) {
+      return null;
+    }
+    return response.getContainerData();
+  }
+
+  public ContainerInfoToken getContainerInfoToken(long containerId)
+      throws IOException {
+    ContainerInfoToken cachedData = encodedTokenCache.getIfPresent(containerId);
+    if (cachedData != null) {
+      return cachedData;
+    }
+    // Cache miss - fetch and store
+    ContainerInfo info = containerOperationClient.getContainer(containerId);
+    String encodeToken = containerOperationClient.getEncodedContainerToken(containerId);
+    cachedData = new ContainerInfoToken(info, encodeToken);
+    encodedTokenCache.put(containerId, cachedData);
+    return cachedData;
+  }
+}
+
+class ReplicaKey {
+  private final DatanodeDetails datanode;
+  private final long containerID;
+
+  ReplicaKey(DatanodeDetails datanode, long containerID) {
+    this.datanode = datanode;
+    this.containerID = containerID;
+  }
+
+  @Override
+  public boolean equals(Object o) {
+    if (this == o) {
+      return true;
+    }
+    if (!(o instanceof ReplicaKey)) {
+      return false;
+    }
+    ReplicaKey key = (ReplicaKey) o;
+    return Objects.equals(datanode, key.datanode) &&
+        Objects.equals(containerID, key.containerID);
+  }
+
+  @Override
+  public int hashCode() {
+    return Objects.hash(datanode, containerID);
+  }
+}
+
+class ContainerInfoToken {
```

Comment:
If we need a helper class for the cache let's make it private within the verifier class instead of package-private.